File: frontend/src/services/alerts.ts
import { apiFetch } from "@/lib/api"
import type { AlertListResponse } from "@/types/alert"

export async function getAlerts(
  params: { case_id?: number; include_dismissed?: boolean } = {}
): Promise<AlertListResponse> {
  const qs = new URLSearchParams()
  if (params.case_id !== undefined) qs.set("case_id", String(params.case_id))
  if (params.include_dismissed) qs.set("include_dismissed", "true")
  const url = `/api/v1/alerts${qs.toString() ? `?${qs}` : ""}`
  const res = await apiFetch(url)
  if (!res.ok) throw new Error("Failed to fetch alerts")
  return res.json()
}

export async function dismissAlert(input: {
  case_id: number
  rule_id: string
  fingerprint?: string | null
  reason?: string
}) {
  const res = await apiFetch("/api/v1/alerts/dismiss", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  })
  if (!res.ok) throw new Error("Failed to dismiss alert")
  return res.json()
}

export async function undismissAlert(caseId: number, ruleId: string) {
  const qs = new URLSearchParams({ case_id: String(caseId), rule_id: ruleId })
  const res = await apiFetch(`/api/v1/alerts/dismiss?${qs}`, { method: "DELETE" })
  if (!res.ok) throw new Error("Failed to restore alert")
  return res.json()
}

File: routes/alerts.py
"""
Case health alert routes.

Computes data-quality / consistency issues across cases and exposes
dismiss / un-dismiss endpoints.
"""

import asyncio
from fastapi.responses import JSONResponse

import db
import auth
from .common import api_error
from .comments import _get_db_user_id


def register_alert_routes(mcp):
    """Register case-health alert routes."""

    @mcp.custom_route("/api/v1/alerts", methods=["GET"])
    async def api_list_alerts(request):
        """List current case-health alerts. Optional ?case_id= and ?include_dismissed=."""
        if err := auth.require_auth(request):
            return err
        case_id_param = request.query_params.get("case_id")
        case_id = int(case_id_param) if case_id_param else None
        include_dismissed = request.query_params.get("include_dismissed", "false").lower() == "true"

        result = await asyncio.to_thread(db.get_case_alerts, case_id, include_dismissed)
        return JSONResponse(result)

    @mcp.custom_route("/api/v1/alerts/dismiss", methods=["POST"])
    async def api_dismiss_alert(request):
        """Dismiss a dismissible alert. Body: {case_id, rule_id, fingerprint?, reason?}."""
        if err := auth.require_auth(request):
            return err
        body = await request.json()
        case_id = body.get("case_id")
        rule_id = body.get("rule_id")
        if not case_id or not rule_id:
            return api_error("case_id and rule_id are required", "MISSING_FIELDS")

        user = auth.get_current_user(request)
        user_id = _get_db_user_id(user)
        try:
            result = await asyncio.to_thread(
                db.dismiss_alert, int(case_id), rule_id,
                body.get("fingerprint"), user_id, body.get("reason"),
            )
        except ValueError as e:
            return api_error(str(e), "INVALID_RULE")
        return JSONResponse(result)

    @mcp.custom_route("/api/v1/alerts/dismiss", methods=["DELETE"])
    async def api_undismiss_alert(request):
        """Un-dismiss an alert. Query params: case_id, rule_id."""
        if err := auth.require_auth(request):
            return err
        case_id = request.query_params.get("case_id")
        rule_id = request.query_params.get("rule_id")
        if not case_id or not rule_id:
            return api_error("case_id and rule_id are required", "MISSING_FIELDS")

        result = await asyncio.to_thread(db.undismiss_alert, int(case_id), rule_id)
        return JSONResponse(result)
